### DISCUSSION TITLE
[@types/react]: ReactNode type definition seems to be too loose

### DISCUSSION THREAD

**Discussion by stefanvonderkrone** [DefinitelyTyped/DefinitelyTyped#55422] [Issues with a @types package]
Hi,
lately I came across a strange type-checking behavior with @types/react.

```ts
import { ReactNode, ReactFragment } from 'react';

const content: ReactNode = { foo: <></> };
const fragment: ReactFragment = { foo: <></> };
```

My understanding is, that this should not pass the type-checker, but it certainly does.
I tracked that issue down to `ReactFragment` which is defined by `type ReactFragment = {} | ReactNodeArray;`. So it seems, the `{}` is causing this issue. But what should be the correct definition for `ReactFragment`.

At the end, TypeScript is happy with the code, but the React runtime will throw an exception when it tries to render the content.

Thank you,
Stefan :)

[a comment by an automated account is not shown]

**Comment by dimitropoulos** [comment]
Could you try to do some git sleuthing and find the commit where the `{}` was introduced so we can try to determine if there was a rationale for doing this?  Usually, there's some reason and removing it may fix one problem but create another in the process.  With that knowledge, we may, of course, end up all deciding on one direction over the other.

**Reply by stefanvonderkrone** [reply]
thank for the reply, I will investigate and report back :)

**Reply by stefanvonderkrone** [reply]
so this seems to be the commit, that introduced the `{}` for `ReactFragment`:
https://github.com/DefinitelyTyped/DefinitelyTyped/commit/57340eca1e3aac68d59cc981b441da834ca38235
But there does not seem to be a hint on why the empty object was chosen. As for the actual return value of `<></>` from 'React.createElement`, it looks like this:
```ts
{
  key: null, // sure, `string | number` as well, of course
  ref: null, // Fragment cannot have refs, I guess
  props: {}, // Fragments also have no props other than `children` 
}
```
So the return-type of `<Fragment />` should be the following?
```ts
type Fragmet = {
  key: string | number | null;
  ref: null;
  props: {
    children?: ReactNode;
  };
}
```
But yeah, you're probably right. A fix of this kind can introduce other problems 🙈

**Reply by stefanvonderkrone** [reply]
here is a very good explanation on why the type definition of `ReactFragment` is an issue:
https://2ality.com/2020/01/typing-objects-typescript.html#excess-property-checks
(see the part for an empty interface)

**Reply by dimitropoulos** [reply]
Ah, ok.  So then it sounds to me like `{}` was used to get around excess property checks.  Now the question remains - why might that be?  Any ideas?

**Reply by stefanvonderkrone** [reply]
I don't have a single one. I wonder if the author had some other `Fragment` in mind. If I recall correctly `React.Fragment` was introduced way after React version 0.13, but I might also be wrong.
I'd suggest I make a PR. Let's see how that turns out

**Reply by dimitropoulos** [reply]
If you make a PR, I'll ask that you track down some larger projects (to plug a personal project, haha, you could find some good candidates on with https://react-ui-roundup.dimitrimitropoulos.com/) and run this change on their codebase and see if it turns up any new errors.

**Reply by stefanvonderkrone** [reply]
that's a good idea. We have a fairly large codebase ourselves. I'll take your candidates for a spin as well.
Thank you.

**Reply by dimitropoulos** [reply]
Sweet.  Let us know how it goes :)

**Reply by lgenzelis** [reply]
This has been fixed in #56210 :)
